### ISSUE TITLE
Feature Request: Enable non-R code exercises

### ISSUE THREAD

**Issue by lukaskawerau** [rstudio/learnr#213]
In normal Rmd documents you can specify SQL-only chunks, based on a database connection via DBI:

````
```{r}
library(DBI)
mydata = dbConnect(RPostgres::Postgres(), 
               dbname = "some_database",
               host = 'localhost',
               port = 5432,
               user = 'postgres',
               password = 'mysecretpassword')
```
````

````
```{sql, connection=mydata}
CREATE TABLE table_name (
  PersonID int,
  LastName varchar(255),
  FirstName varchar(255)
);
```
````

However, this does not work with `learnr`. I would like to be able to do the same in an exercise so that students can directly submit SQL statements:

````
```{sql, connection=mydata, exercise = TRUE}

```
````

Where students then would have to enter SQL code into the exercise.

**Comment by schloerke** [comment]
I'm going to leave this open as it's a good idea.

It's going to be a while as the correct infrastructure is not there.

**Comment by coatless** [comment]
@schloerke let's generalize to all of [R Markdown's supported engines](https://bookdown.org/yihui/rmarkdown/language-engines.html)? 

Outside of SQL, it would be useful to have Python.

**Comment by zoews** [comment]
Hi, @coatless @schloerke should I assume that Python isn't working with learnr at this time? Is there any workaround with e.g. reticulate? If not, I would be interested in getting involved with adding this functionality as it would be extremely helpful for a work project (slash helpful in general!)

**Comment by schloerke** [comment]
@zoews 
> should I assume that Python isn't working with learnr at this time? 

Correct

> Is there any workaround with e.g. reticulate?

I would like to use knitr's engine to process the information.  R Markdown already [has support for python](https://rstudio.github.io/reticulate/articles/r_markdown.html).

This would allow for all existing languages that can produce output to work as expected.  (Persistent DB connections or shiny applications will not be possible.  Only code that evaluates and returns is possible in the current execution model.)


### Recipe
[`evaluate_exercise`](https://github.com/rstudio/learnr/blob/5571b39c56888aabf1eccc5dc053883a32a1328d/R/exercise.R#L106) is where the execution of the user exercise happens. 

I'm guessing that (as of writing this comment)
* the [`knitr::spin`](https://github.com/rstudio/learnr/blob/5571b39c56888aabf1eccc5dc053883a32a1328d/R/exercise.R#L221-L224) call will have to have a different function call when working with different languages.
* the code pertaining to [`last_value` and `learnr_output_handler`](https://github.com/rstudio/learnr/blob/5571b39c56888aabf1eccc5dc053883a32a1328d/R/exercise.R#L234-L267) will need to not be executed
* the `checker` functions ([pre](https://github.com/rstudio/learnr/blob/5571b39c56888aabf1eccc5dc053883a32a1328d/R/exercise.R#L142-L151) and [post](https://github.com/rstudio/learnr/blob/5571b39c56888aabf1eccc5dc053883a32a1328d/R/exercise.R#L335)) might need an extra argument of `language`.

Debugging will be hard with a temp file being created each time.  I recommend changing [this line](https://github.com/rstudio/learnr/blob/5571b39c56888aabf1eccc5dc053883a32a1328d/R/exercise.R#L174) to a local working path folder.

--------------------

@zoews For implementation ideas/discussion/troubleshooting, please open a PR. 😄

**Comment by braunsb** [comment]
Just leaving my vote: If we had this feature, we would be using learnr all over the organization!

**Comment by TonyFly3000** [comment]
would love to have following Feature

```{sql, connection=mydata, exercise = TRUE}

```

**Comment by JosiahParry** [comment]
Just want to throw in my vote for this as well :)

**Comment by schloerke** [comment]
@nischalshrestha has re-written how exercises are calculated in #390 (just merged to master). Thank you @zoews for getting the conversation started and showing what is possible.

Currently, no UI changes have been merged into master from #310. Working on this in #397

If you could test this locally, that would be great! If you run into any issues, please comment on #397, as I will be making followup adjustments there. Thank you!

<details>
<summary><a href="https://github.com/rstudio/learnr/pull/397/files#diff-dc1f72612c6759cda193132f17e7add7">Example polyglot tutorial</a></summary>

````
---
title: "Multi-language exercises"
author: "Zoë Wilkinson Saldaña"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include = FALSE}
library(learnr)

# altered from https://datacarpentry.org/R-ecology-lesson/05-r-and-databases.html
if (!file.exists("portal_mammals.sqlite")) {
  download.file(url = "https://ndownloader.figshare.com/files/2292171",
                destfile = "portal_mammals.sqlite", mode = "wb")
}
mammals <- DBI::dbConnect(RSQLite::SQLite(), "portal_mammals.sqlite")
```

### Python or R?

The following two code exercises run language agnostic code.  It will finally evaluate to either 5 or 10.

```{r r-not-python, exercise=TRUE, exercise.lines = 5}
# if running in R, this should return 10
# if running in Python, this should return FALSE and 5
x = 5
x <-10
print(x)
```

```{python python-not-r, exercise=TRUE, exercise.lines = 5}
# if running in R, this should return 10
# if running in Python, this should return FALSE and 5
x = 5
x <-10
print(x)
```

### Bash

Execute exercises in bash

```{bash bash-not-r, exercise = TRUE}
echo "hello world"
date
```

### SQL

Use a SQL connection defined in a setup chunk.

```{sql sql-not-r, exercise = TRUE, connection="mammals"}
SELECT *
FROM `surveys`
LIMIT 10
```
````

</details>

<img width="1553" alt="Screen Shot 2020-07-14 at 11 24 13 AM" src="https://user-images.githubusercontent.com/93231/87444265-91498d80-c5c4-11ea-8bb4-d9784462559c.png">

-----------------

Closing as fixed in #390
